### DISCUSSION TITLE
理解 Go 语言 defer 关键字的原理 · /golang-defer

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#560] [Announcements]
https://draveness.me/golang-defer

**Comment by auxten** [comment]
恭喜博主成为 k8s & istio member

**Comment by draveness** [comment]
> 恭喜博主成为 k8s & istio member

...... 🤣

**Comment by panxue** [comment]
非常感谢~

**Comment by zerozh** [comment]
```
func main() {
	for i := 0; i < 5; i++ {
		defer fmt.Println(i)
	}
}

$ go run main.go
5
4
3
2
1
```
的输出结果应为 4,3,2,1,0

**Comment by draveness** [comment]
> ```
> func main() {
> 	for i := 0; i < 5; i++ {
> 		defer fmt.Println(i)
> 	}
> }
> 
> $ go run main.go
> 5
> 4
> 3
> 2
> 1
> ```
> 的输出结果应为 4,3,2,1,0

已经修复了，非常感谢

**Comment by Huweicai** [comment]
恭喜大佬成为 k8s & istio member！！！

**Comment by stubbornTanzhe** [comment]
赞

**Comment by g10guang** [comment]
简单总结就是 goroutine 结构体中会保存 defer 调用链，函数 return 时候会判断 sp 和 goroutine defer 链中的 sp 和当前 sp 关系，如果符合条件就执行 defer 结构体中保存的函数指针，参数也是通过 defer 结构体保存的传递过去的。
所以 defer 函数传递的是值传递。

**Comment by wwlwwww** [comment]
> @g10guang
> 简单总结就是 goroutine 结构体中会保存 defer 调用链，函数 return 时候会判断 sp 和 goroutine defer 链中的 sp 和当前 sp 关系，如果符合条件就执行 defer 结构体中保存的函数指针，参数也是通过 defer 结构体保存的传递过去的。
> 所以 defer 函数传递的是值传递。

“函数 return 时候会判断 sp 和 goroutine defer 链中的 sp 和当前 sp 关系”，没在文中看到比较的地方呢

**Comment by wwlwwww** [comment]
deferreturn()函数是不是少了一部分：
```golang
sp := getcallersp()
if d.sp != sp {
      return
}
```

**Comment by draveness** [comment]
> @WMLHUST
> deferreturn()函数是不是少了一部分：
> ```golang
> sp := getcallersp()
> if d.sp != sp {
>       return
> }
> ```
> 

对，代码时经过删减的

**Comment by cloudjjcc** [comment]
> 创建延迟调用
runtime.deferproc 会为 defer 创建一个新的 runtime._defer 结构体、设置它的函数指针 fn、程序计数器 sp 和栈指针 sp 并将相关的参数拷贝到相邻的内存空间中

这里有笔误，程序计数器应该是pc

**Comment by draveness** [comment]
> > 创建延迟调用
> > runtime.deferproc 会为 defer 创建一个新的 runtime._defer 结构体、设置它的函数指针 fn、程序计数器 sp 和栈指针 sp 并将相关的参数拷贝到相邻的内存空间中
> 
> 这里有笔误，程序计数器应该是pc

修复了，感谢

**Comment by gabryzhong** [comment]
感觉一群建筑设计师在学沥青制作，还觉得沥青师傅很牛逼

**Comment by draveness** [comment]
> 感觉一群建筑设计师在学沥青制作，还觉得沥青师傅很牛逼

@gabryzhong 先不说这个类比对错吧，沥青师傅确实很牛逼

**Comment by bzd111** [comment]
pruntime.deferproc] => [

**Comment by yanjinbin** [comment]
发现大佬的博客 都是精华啊  有学习了 defer调用过程

**Comment by xlionet** [comment]
hi，大佬，在阅读go defer 源码中，有一个疑惑。
在SSA构建中的call 函数里
```go
func (s *state) call(n *Node, k callKind) *ssa.Value {
....

var call *ssa.Value
	if k == callDeferStack {
...
		s.vars[&memVar] = s.newValue1A(ssa.OpVarDef, types.TypeMem, d, s.mem())
		addr := s.addr(d, false)

		// Must match reflect.go:deferstruct and src/runtime/runtime2.go:_defer.
		// 0: siz
		s.store(types.Types[TUINT32],
			s.newValue1I(ssa.OpOffPtr, types.Types[TUINT32].PtrTo(), t.FieldOff(0), addr),
			s.constInt32(types.Types[TUINT32], int32(stksize)))
		// 1: started, set in deferprocStack
		// 2: heap, set in deferprocStack
		// 3: sp, set in deferprocStack
		// 4: pc, set in deferprocStack
		// 5: fn
		s.store(closure.Type,
			s.newValue1I(ssa.OpOffPtr, closure.Type.PtrTo(), t.FieldOff(5), addr),
			closure)
```

看这个store函数似乎是想去初始化新分配defer 空间，但是store 函数的实现让我很迷惑，

```go
func (s *state) store(t *types.Type, dst, val *ssa.Value) {
	s.vars[&memVar] = s.newValue3A(ssa.OpStore, types.TypeMem, t, dst, val, s.mem())
}
```
memVar 这个是个全局对象，那每次store的时候 ，都是往s.vars 的map的同一个key 去set 值吗，这样不是会覆盖吗

**Comment by draveness** [comment]
> hi，大佬，在阅读go defer 源码中，有一个疑惑。
> 在SSA构建中的call 函数里
> 
> ```go
> func (s *state) call(n *Node, k callKind) *ssa.Value {
> ....
> 
> var call *ssa.Value
> 	if k == callDeferStack {
> ...
> 		s.vars[&memVar] = s.newValue1A(ssa.OpVarDef, types.TypeMem, d, s.mem())
> 		addr := s.addr(d, false)
> 
> 		// Must match reflect.go:deferstruct and src/runtime/runtime2.go:_defer.
> 		// 0: siz
> 		s.store(types.Types[TUINT32],
> 			s.newValue1I(ssa.OpOffPtr, types.Types[TUINT32].PtrTo(), t.FieldOff(0), addr),
> 			s.constInt32(types.Types[TUINT32], int32(stksize)))
> 		// 1: started, set in deferprocStack
> 		// 2: heap, set in deferprocStack
> 		// 3: sp, set in deferprocStack
> 		// 4: pc, set in deferprocStack
> 		// 5: fn
> 		s.store(closure.Type,
> 			s.newValue1I(ssa.OpOffPtr, closure.Type.PtrTo(), t.FieldOff(5), addr),
> 			closure)
> ```
> 
> 看这个store函数似乎是想去初始化新分配defer 空间，但是store 函数的实现让我很迷惑，
> 
> ```go
> func (s *state) store(t *types.Type, dst, val *ssa.Value) {
> 	s.vars[&memVar] = s.newValue3A(ssa.OpStore, types.TypeMem, t, dst, val, s.mem())
> }
> ```
> 
> memVar 这个是个全局对象，那每次store的时候 ，都是往s.vars 的map的同一个key 去set 值吗，这样不是会覆盖吗

这个东西就是一个 `state` 中的临时变量

+ [`state.mem`](https://github.com/golang/go/blob/4d5bb9c60905b162da8b767a8a133f6b4edcaa65/src/cmd/compile/internal/gc/ssa.go#L5701) 这个方法会得到 memVar 中的值
+ [`state.store`](https://github.com/golang/go/blob/4d5bb9c60905b162da8b767a8a133f6b4edcaa65/src/cmd/compile/internal/gc/ssa.go#L996) 会拿到之前临时变量中的值进行计算并覆盖 memVar

**Comment by xlionet** [comment]
> @draveness
> > hi，大佬，在阅读go defer 源码中，有一个疑惑。
> > 在SSA构建中的call 函数里
> > 
> > ```go
> > func (s *state) call(n *Node, k callKind) *ssa.Value {
> > ....
> > 
> > var call *ssa.Value
> > 	if k == callDeferStack {
> > ...
> > 		s.vars[&memVar] = s.newValue1A(ssa.OpVarDef, types.TypeMem, d, s.mem())
> > 		addr := s.addr(d, false)
> > 
> > 		// Must match reflect.go:deferstruct and src/runtime/runtime2.go:_defer.
> > 		// 0: siz
> > 		s.store(types.Types[TUINT32],
> > 			s.newValue1I(ssa.OpOffPtr, types.Types[TUINT32].PtrTo(), t.FieldOff(0), addr),
> > 			s.constInt32(types.Types[TUINT32], int32(stksize)))
> > 		// 1: started, set in deferprocStack
> > 		// 2: heap, set in deferprocStack
> > 		// 3: sp, set in deferprocStack
> > 		// 4: pc, set in deferprocStack
> > 		// 5: fn
> > 		s.store(closure.Type,
> > 			s.newValue1I(ssa.OpOffPtr, closure.Type.PtrTo(), t.FieldOff(5), addr),
> > 			closure)
> > ```
> > 
> > 看这个store函数似乎是想去初始化新分配defer 空间，但是store 函数的实现让我很迷惑，
> > 
> > ```go
> > func (s *state) store(t *types.Type, dst, val *ssa.Value) {
> > 	s.vars[&memVar] = s.newValue3A(ssa.OpStore, types.TypeMem, t, dst, val, s.mem())
> > }
> > ```
> > 
> > memVar 这个是个全局对象，那每次store的时候 ，都是往s.vars 的map的同一个key 去set 值吗，这样不是会覆盖吗
> 
> 这个东西就是一个 `state` 中的临时变量
> 
> + [`state.mem`](https://github.com/golang/go/blob/4d5bb9c60905b162da8b767a8a133f6b4edcaa65/src/cmd/compile/internal/gc/ssa.go#L5701) 这个方法会得到 memVar 中的值
> + [`state.store`](https://github.com/golang/go/blob/4d5bb9c60905b162da8b767a8a133f6b4edcaa65/src/cmd/compile/internal/gc/ssa.go#L996) 会拿到之前临时变量中的值进行计算并覆盖 memVar

看懂了，感谢~~

**Comment by hzldex** [comment]
5.3.1的 预计算参数 部分的最后这段

>虽然调用 defer 关键字时也使用值传递，但是因为拷贝的是函数指针，所以 time.Since(startedAt) 会在 main 函数**执行**前被调用并打印出符合预期的结果

此处的 执行 似乎是笔误了，应该是 返回 ？


---

*2020-06-10 UPDATES: 已修复*

```diff
-所以 time.Since(startedAt) 会在 main 函数执行前被调用并打印出符合预期的结果
+所以 time.Since(startedAt) 会在 main 函数返回前被调用并打印出符合预期的结果
```

**Comment by huxiaoyugo** [comment]
> @gabryzhong
> 感觉一群建筑设计师在学沥青制作，还觉得沥青师傅很牛逼

大佬牛逼

**Comment by Northward712** [comment]
了解了大概的执行方式，但是源代码真的是一部分看不懂

**Comment by Imfan** [comment]
>在默认情况下，我们可以看到 Go 语言中 runtime._defer 结构体都会在堆上分配，如果我们能够将部分结构体分配到堆上就可以节约内存分配带来的额外开销

应该是  将部分结构体分配到栈上 吧


---

*2021-01-19 UPDATES: 已修复*

**Comment by draveness** [comment]
> @E-c-h-o-Go 了解了大概的执行方式，但是源代码真的是一部分看不懂

不敢兴趣就看看顶层设计，感兴趣就多看看

**Comment by dillanzhou** [comment]
最大的疑问是：为啥deferreturn要使用jmpdefer这种方式实现？为什么不能在deferreturn中直接写个循环调用defer链表上的函数？

**Comment by draveness** [comment]
> 最大的疑问是：为啥deferreturn要使用jmpdefer这种方式实现？为什么不能在deferreturn中直接写个循环调用defer链表上的函数？

jmpdefer 利用了函数的调用惯例，按照调用管理在栈上做了一些操作，实现了特殊的跳转逻辑，可以详细看一下

**Comment by czyt** [comment]
这个[位置](https://draveness.me/golang/docs/part2-foundation/ch05-keyword/golang-defer/#535-%e5%bc%80%e6%94%be%e7%bc%96%e7%a0%81)应该是开放编码 而不是开发编码

---

*2021-03-09 UPDATES: 已修复*

**Comment by oryxtc** [comment]
> @gabryzhong
> 感觉一群建筑设计师在学沥青制作，还觉得沥青师傅很牛逼

我一个初学go的看了文章都觉得受益匪浅,能避免很多坑 和知道坑的原理...感谢博主!

**Comment by fwan922** [comment]
我真是菜，看不懂

**Comment by draveness** [comment]
> @fwan922 我真是菜，看不懂

可能是写的还不够易懂，这部分很绕

**Comment by herac1es** [comment]
编译期将 defer 关键字被转换 runtime.deferproc 并在调用 defer 关键字的函数返回之前插入 runtime.deferreturn；

大佬，感觉这句话的将和被不太通顺

---

*2021-04-06 UPDATES: 已修复*

**Comment by jwrookie** [comment]
很多 defer 语句都可以在编译期间判断是否被执行，如果函数中的 defer 语句都会在编译期间确定，中间代码生成阶段就会直接调用 cmd/compile/internal/gc.state.openDeferExit 在函数返回前生成判断 deferBits 的代码，也就是上述伪代码中的后半部分。

不过当程序遇到运行时才能判断的条件语句时，我们仍然需要由运行时的 runtime.deferreturn 决定是否执行 defer 关键字：

```
func main() {
	defer fmt.Println(1)
}
```

go tool compile -S  输出的汇编还是会有CALL    runtime.deferreturn(SB)
go version 1.16

**Comment by jackzhou121** [comment]
请问defer结构体中的为啥需要保存sp和pc值，因为执行defer本身就是函数调用，就有压栈的过程，所以不明白为啥defer需要另外两个变量保存sp和pc

**Comment by flyskea** [comment]
> @jackzhou121
> 请问defer结构体中的为啥需要保存sp和pc值，因为执行defer本身就是函数调用，就有压栈的过程，所以不明白为啥defer需要另外两个变量保存sp和pc

sp用于判断defer是否是当前函数注册的，pc是跳转回defer之前函数执行时的下一条指令，因为panic要去执行defer

**Comment by DeepBlue98** [comment]
6666

**Comment by gongqij** [comment]
go1.18以后deferproc咋不传参数大小siz了。
// Create a new deferred function fn, which has no arguments and results.
// The compiler turns a defer statement into a call to this.
func deferproc(fn func()) {...}

**Comment by NingYuanxiang** [comment]
为什么defer在函数体中最多执行一次时会将结构体分配到栈上并调用? 为什么不能全部分配到栈上调用？

**Comment by Javcurry** [comment]
能不能配合调用链的图，讲起来会更清晰，纯文字看起来很绕

**Comment by Javcurry** [comment]
文章中很多源码基本是看不懂的，很烦

**Reply by draveness** [reply]
不用完全看懂，大概理解意思就可以了，defer 相关的实现确实比较复杂，涉及底层很多要多想一下，我之前在研究这些内容的时候也花了很长时间

**Reply by Huweicai** [reply]
是的是的，大佬说的有道理

**Comment by rjxznb** [comment]
“最后调用的 runtime.return0 是唯一一个不会触发延迟调用的函数，它可以避免递归 runtime.deferreturn 的递归调用。” 请问在5.3.4 原文中这句话中的 runtime.return0 是唯一一个不会触发延迟调用的函数是什么意思呢？
